text/0020-unique-record-keys.md

```diff
@@ -0,0 +1,103 @@
+In the case of `context`, the error would need to be caught when constructing the `Context` object.
+In other cases, e.g. if evaluating a Cedar subexpression produces a record with duplicate keys, the user would see an evaluation error. (But see "Unresolved questions" below.)
+
+I don't believe this change would have a significant effect on validation or width/depth subtyping of records, but please correct me if I'm wrong.
+
+## Drawbacks
+
+This is a breaking change, not just to the Cedar API surface, but to the Cedar language (policy syntax) itself.
+That's a significant drawback not to be taken lightly.
+In this case, it's mitigated somewhat by the fact that we're turning previously-valid policies into errors, not changing them into a silently different behavior.
+It's also mitigated by the fact that most Cedar users are probably (citation needed) not writing policies in the class that will have a behavior change -- i.e., policies with duplicate keys in record literals.
+
+## Alternatives
+
+The main alternative is doing nothing, which of course has the reverse pros/cons of the proposal as explained in "Motivation" and "Drawbacks".
+
+Another alternative, which would address the first motivation and not the second, would be to change the EST format so that it supports record literals with duplicate keys.
+I believe this would require a breaking change to the EST format, and record literals would not be expressible as JSON maps, which is unfortunate because I believe that's the most natural/ergonomic.
+(Devil's advocate: we could maybe find a way to support JSON maps in the common case with some escape syntax for expressing record literals that have duplicate keys. If we want to take this alternative seriously, I can investigate this.)
+
+## Unresolved questions
+
+### Evaluation errors for duplicate keys
+Is it actually possible in today's Cedar, to have duplicate keys in a record value other than a record literal or `context`? I.e., to have duplicate keys in a record value produced by evaluating some subexpression (other than a record literal or the `context` variable)?
+Cedar has no way to augment or union record values, and does not allow computed keys in record literals, so off the top of my head the only way to do this would be via `GetAttr`, which would imply duplicate keys in the entity store, and we can/should disallow that as well.
+If I'm not missing something and there actually is no other way, then we can catch all duplicate-keys errors prior to calling `is_authorized()`, and we don't need a new category of evaluation error.
+
+### Order-independence for record keys
+At first glance, one would hope that this change would be sufficient to ensure that the order of keys in a record literal (or more generally, any record value) does not matter.
+However, I don't think that's quite true -- if keys are mapped to expressions with errors, the order does matter, because only the first error encountered will be returned.
```

Comment:
The order of evaluation still matters, even without duplicates.  This is true for both set and record literals, and it's indeed due to error behavior.